org.eclipse.scanning.sequencer/src/org/eclipse/scanning/sequencer/LocationManager.java

```diff
@@ -0,0 +1,184 @@
+		this.bean    = bean;
+		this.model   = model;
+		this.manager = manager;
+		manager.addDevices(this);
+	}
+	
+	public int getOuterSize() {
+		return outerSize;
+	}
+	public void setOuterSize(int outerSize) {
+		this.outerSize = outerSize;
+	}
+	public int getOuterCount() {
+		return outerCount;
+	}
+	public void setOuterCount(int outerCount) {
+		this.outerCount = outerCount;
+	}
+	public int getInnerSize() {
+		return innerSize;
+	}
+	public void setInnerSize(int innerSize) {
+		this.innerSize = innerSize;
+	}
+	public int getTotalSize() {
+		return totalSize;
+	}
+	public void setTotalSize(int totalSize) {
+		this.totalSize = totalSize;
+	}
+	public int getStepNumber() {
+		return stepNumber;
+	}
+	public void setStepNumber(int stepNumber) {
+		this.stepNumber = stepNumber;
+	}
+	
+	/**
+	 * Called during the scan to increment counts.
+	 */
+	@PointEnd
+	public void increment() {
+    	outerCount++;
+		stepNumber+=Math.max(innerSize, 1);
+	}
+
+	/**
+	 * Method used to generate an iterator for the scan.
+	 * It sets counts which are incremented during the scan.
+	 * 
+	 * @return
+	 * @throws ScanningException
+	 */
+	public Iterator<IPosition> createPositionIterator() throws ScanningException {
+		
+		CompoundModel<?> cmodel = bean.getScanRequest()!=null ? bean.getScanRequest().getCompoundModel() : null;
+		SubscanModerator moderator = new SubscanModerator(model.getPositionIterable(), cmodel, model.getDetectors(), ServiceHolder.getGeneratorService());
+		manager.addContext(moderator);
+		
+		try {
+			stepNumber = 0;
+			outerSize  = getEstimatedSize(moderator.getOuterIterable());
+			innerSize  = getEstimatedSize(moderator.getInnerIterable());
+			totalSize  = getEstimatedSize(model.getPositionIterable());
+		} catch (GeneratorException se) {
+			throw new ScanningException("Cannot create the position iterator!", se);
+		}
+
+		return moderator.getOuterIterable().iterator();
+	}
+
+
+	private int getEstimatedSize(Iterable<IPosition> gen) throws GeneratorException {
+		
+		int size=0;
+		if (gen instanceof IDeviceDependentIterable) {
+			size = ((IDeviceDependentIterable)gen).size();
+			
+		} else if (gen instanceof IPointGenerator<?>) {
+			size = ((IPointGenerator<?>)gen).size();
+			
+		} else if (gen!=null) {
+		    for (IPosition unused : gen) size++; // Fast even for large stuff providing they do not check hardware on the next() call.
+		}
+		return size;   		
+	}
+
+	/**
+	 * Seek within the iterator for the given location.
+	 * @param location
+	 * @param iterator
+	 * @return null if position not found.
+	 */
+	public IPosition seek(int location, Iterator<IPosition> iterator) {
+		/*
+		 * IMPORTANT We do not keep the positions in memory because there can be millions.
+		 * Running over them is fast however.
+		 */
+		while(iterator.hasNext()) {
+			IPosition pos = iterator.next();
+        	pos.setStepIndex(stepNumber);
+			if (stepNumber == location) return pos;

```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') At most one statement is allowed per line, but 2 statements were found on this line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00122)